common/functions/src/scalars/strings/substring.rs

```diff
@@ -14,235 +14,97 @@
 use crate::scalars::function_factory::FunctionFeatures;
-use crate::scalars::Function;
+use crate::scalars::Function2;
+use crate::scalars::Function2Description;
 
 #[derive(Clone)]
 pub struct SubstringFunction {
     display_name: String,
 }
 
 impl SubstringFunction {
-    pub fn try_create(display_name: &str) -> Result<Box<dyn Function>> {
+    pub fn try_create(display_name: &str) -> Result<Box<dyn Function2>> {
         Ok(Box::new(SubstringFunction {
             display_name: display_name.to_string(),
         }))
     }
 
-    pub fn desc() -> FunctionDescription {
-        FunctionDescription::creator(Box::new(Self::try_create)).features(
+    pub fn desc() -> Function2Description {
+        Function2Description::creator(Box::new(Self::try_create)).features(
             FunctionFeatures::default()
                 .deterministic()
                 .variadic_arguments(2, 3),
         )
     }
 }
 
-impl Function for SubstringFunction {
+impl Function2 for SubstringFunction {
     fn name(&self) -> &str {
         &*self.display_name
     }
 
-    fn return_type(&self, args: &[DataTypeAndNullable]) -> Result<DataTypeAndNullable> {
-        if !args[0].is_string() && !args[0].is_null() {
+    fn return_type(&self, args: &[&DataTypePtr]) -> Result<DataTypePtr> {
+        if !args[0].data_type_id().is_string() && !args[0].data_type_id().is_null() {
             return Err(ErrorCode::IllegalDataType(format!(
                 "Expected string or null, but got {}",
-                args[0]
+                args[0].data_type_id()
             )));
         }
-        if !args[1].is_integer() && !args[1].is_null() {
+
+        if !args[1].data_type_id().is_integer() && !args[1].data_type_id().is_null() {
             return Err(ErrorCode::IllegalDataType(format!(
                 "Expected integer or string or null, but got {}",
-                args[1]
+                args[1].data_type_id()
             )));
         }
-        if args.len() > 2 && !args[2].is_integer() && !args[2].is_null() {
+
+        if args.len() > 2
+            && !args[2].data_type_id().is_integer()
+            && !args[2].data_type_id().is_null()
+        {
             return Err(ErrorCode::IllegalDataType(format!(
-                "Expected integer or null, but got {}",
-                args[2]
+                "Expected integer or string or null, but got {}",
+                args[2].data_type_id()
             )));
         }
-        let nullable = args.iter().any(|arg| arg.is_nullable());
-        let dt = DataType::String;
-        Ok(DataTypeAndNullable::create(&dt, nullable))
+
+        Ok(StringType::arc())
     }
 
-    fn eval(&self, columns: &DataColumnsWithField, input_rows: usize) -> Result<DataColumn> {
-        let s_column = columns[0].column().cast_with_type(&DataType::String)?;
-        let p_column = columns[1].column().cast_with_type(&DataType::Int64)?;
+    fn eval(&self, columns: &ColumnsWithField, input_rows: usize) -> Result<ColumnRef> {
```

Comment:
thanks for column viewer, this complicated match could be avoid cc @sundy-li

Reply:
Great!  But I suggest merging main into this pr to use the latest version of `Viewer `.

Reply:
let viewer = i16::try_create_viewer(&column)?;

    assert_eq!(viewer.size(), 1024);
    assert!(!viewer.null_at(0));
    assert!(!viewer.null_at(99));

    for (idx, value) in viewer.iter().enumerate() {
        if viewer.null_at(idx) {
            assert_eq!(value, 0);
        } else {
            assert_eq!(value, 123i16);
        }
    }

Reply:
resolved